source/module_base/test/complexarray_test.cpp

```diff
@@ -45,125 +50,416 @@
+*   - functon abs2(), return the sum of squares of all elements 
+*   - function dot()
+*   - function scale_accumulate()
+*   - function scaled_sum()
+*   - function point_mult()
 */
 
+//compare two complex by using EXPECT_DOUBLE_EQ()
+void EXPECT_COMPLEX_EQUAL(const std::complex<double>& a,const std::complex<double>& b)
+{
+    EXPECT_DOUBLE_EQ(a.real(),b.real());
+    EXPECT_DOUBLE_EQ(a.imag(),b.imag());
+}
 
 class ComplexArray_test : public testing::Test
 {
     protected:
         ModuleBase::ComplexArray a2,a4,b2,b4,c2,c4,d2;
-        ModuleBase::ComplexArray a2_plus_b2,a2_minus_b2,a2_mult_b2,a2_mult_3;
         std::complex<double> com1 {1.0,2.0};
         std::complex<double> com2 {3.0,4.0};
+        std::complex<double> com3 {-2.0,-3.0};
+        std::complex<double> comzero {0.0,0.0};
 
         void SetUp()
         {
             a2 = ModuleBase::ComplexArray(2,1,1,1);  // if define this class as a matrix
             b2 = ModuleBase::ComplexArray(2,1,1,1);  // of 4 dimesions,
             c2 = ModuleBase::ComplexArray(2,1,1,1);  // is a2 +/-/* d2 allowed or not???
             d2 = ModuleBase::ComplexArray(1,2,1,1);  // it does not matter, this situation will not appear in ABACUS
-            a2_plus_b2 = ModuleBase::ComplexArray(2,1,1,1);
-            a2_minus_b2 = ModuleBase::ComplexArray(2,1,1,1);
-            a2_mult_b2 = ModuleBase::ComplexArray(2,1,1,1);
-            a2_mult_3 = ModuleBase::ComplexArray(2,1,1,1);
             a4 = ModuleBase::ComplexArray(2,2,1,1);
             b4 = ModuleBase::ComplexArray(2,2,1,1);
             c4 = ModuleBase::ComplexArray(2,2,1,1);
-            a2 = com1;
-            b2 = com2;
-            a4 = com1;
-            b4 = com2;
-            d2 = com1;
-            a2_plus_b2 = com1 + com2;
-            a2_minus_b2 = com1 - com2;
-            a2_mult_b2 = com1 * com2;
-            a2_mult_3 = com1 * 3.0;
         }
 
 };
 
-TEST_F(ComplexArray_test,operator_equal)
+
+TEST(ComplexArray, constructor_bnd1234_test)
+{
+    ModuleBase::ComplexArray a(1,2,3,4);
+    ASSERT_EQ(a.getSize(),24);
+    ASSERT_EQ(a.getBound1(),1);
+    ASSERT_EQ(a.getBound2(),2);
+    ASSERT_EQ(a.getBound3(),3);
+    ASSERT_EQ(a.getBound4(),4);
+    /*
+    ASSERT_DEATH(ModuleBase::ComplexArray a(0,2,3,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,0,3,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,2,0,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,2,3,0),"");
+    */
+}
+
+TEST(ComplexArray, constructor_copy_test)
+{
+    ModuleBase::ComplexArray a(1,2,3,4);
+    ModuleBase::ComplexArray b(a);
+    ASSERT_EQ(b.getSize(),24);
+    ASSERT_EQ(b.getBound1(),1);
+    ASSERT_EQ(b.getBound2(),2);
+    ASSERT_EQ(b.getBound3(),3);
+    ASSERT_EQ(b.getBound4(),4);
+       
+    ASSERT_EQ(a.getSize(),24);
+    ASSERT_EQ(a.getBound1(),1);
+    ASSERT_EQ(a.getBound2(),2);
+    ASSERT_EQ(a.getBound3(),3);
+    ASSERT_EQ(a.getBound4(),4);
+
+}
+
+TEST(ComplexArray, constructor_rvalue_test)
```

Comment:
We're already in test, do we need to emphasize in `constructor_rvalue_test`?

Reply:
I think this is not critical, and is just a habit, and will not cause any ambiguity or misunderstanding.